src/manager/manager.c

```diff
@@ -252,9 +254,191 @@ static int manager_method_ping(sd_bus_message *m, UNUSED void *userdata, UNUSED
+        int r = sd_bus_message_open_container(reply, SD_BUS_TYPE_ARRAY, "(sssssssouso)");
+        if (r < 0) {
+                return r;
+        }
+
+        for (int i = 0; i < req->n_sub; i++) {
+                const char *node_name = req->sub[i].node->name;
+                sd_bus_message *m = req->sub[i].m;
+                if (m == NULL) {
+                        continue;
+                }
+
+                r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "(ssssssouso)");
+                if (r < 0) {
+                        return r;
+                }
+
+                while (sd_bus_message_at_end(m, false) == 0) {
+                        r = sd_bus_message_open_container(reply, SD_BUS_TYPE_STRUCT, "sssssssouso");
+                        if (r < 0) {
+                                return r;
+                        }
+                        r = sd_bus_message_enter_container(m, SD_BUS_TYPE_STRUCT, "ssssssouso");
+                        if (r < 0) {
+                                return r;
+                        }
+
+                        r = sd_bus_message_append(reply, "s", node_name);
+                        if (r < 0) {
+                                return r;
+                        }
+                        r = sd_bus_message_copy(reply, m, true);
+                        if (r < 0) {
+                                return r;
+                        }
+
+                        r = sd_bus_message_close_container(reply);
+                        if (r < 0) {
+                                return r;
+                        }
+                        r = sd_bus_message_exit_container(m);
+                        if (r < 0) {
+                                return r;
+                        }
+                }
+
+                r = sd_bus_message_exit_container(m);
+                if (r < 0) {
+                        return r;
+                }
+        }
+
+        r = sd_bus_message_close_container(reply);
+        if (r < 0) {
+                return r;
+        }
+
+        return 0;
+}
+
+
+static void manager_method_list_units_done(ListUnitRequest *req) {
+        /* All sub-requests are done, collect results and freee when done */
+        _cleanup_list_unit_request_ ListUnitRequest *free_me = req;
+
+        _cleanup_sd_bus_message_ sd_bus_message *reply = NULL;
+        int r = sd_bus_message_new_method_return(req->request_message, &reply);
+        if (r >= 0) {
+                r = manager_method_list_units_encode_reply(req, reply);
+        }
+        if (r < 0) {
+                sd_bus_reply_method_errorf(req->request_message, SD_BUS_ERROR_FAILED, "Internal error");
+                return;
+        }
+
+        r = sd_bus_message_send(reply);
+        if (r < 0) {
+                fprintf(stderr, "Failed to send reply: %s\n", strerror(-r));
+                return;
+        }
+}
+
+static void manager_method_list_units_maybe_done(ListUnitRequest *req) {
```

Comment:
Maybe rename this ```manager_method_check_if_list_units_done()```.

Reply:
I think that primarily makes it longer, not necessary easier to understand.